File: tools/metagen/tool/src/com/cpgf/metagen/codewriter/CodeWriter.java
package com.cpgf.metagen.codewriter;

public class CodeWriter {
	private int indent;
	private String text;
	
	public CodeWriter() {
		this.indent = 0;
		this.text = "";
	}

	public void incIndent() {
		++this.indent;
	}

	public void decIndent() {
		--this.indent;
		if(this.indent < 0) {
			this.indent = 0;
		}
	}
	
	public void out(String s) {
		String indentText = "";
		for(int i = 0; i < this.indent; ++i) {
			indentText = indentText + "    "; 
		}

		if(this.text.length() > 0 && this.text.lastIndexOf('\n') == this.text.length() - 1) {
			this.text = this.text + indentText;
		}

		this.text = this.text + s;
	}
	
	public String getText() {
		return this.text;
	}

}

File: tools/cmetagen/src/codewriter/codewriter.cpp
#include "codewriter.h"

namespace metagen {


CodeWriter::CodeWriter()
	: indent(0)
{
}

CodeWriter::~CodeWriter()
{
}

std::string CodeWriter::getIndentText() const
{
	std::string indentText;
	for(int i = 0; i < this->indent; ++i) {
		indentText = indentText + "    "; 
	}
	
	return indentText;
}

void CodeWriter::incIndent()
{
	++this->indent;
}

void CodeWriter::decIndent()
{
	if(this->indent > 0) {
		--this->indent;
	}
}

void CodeWriter::incIndent(int level)
{
	this->indent += level;
}

void CodeWriter::decIndent(int level)
{
	this->indent -= level;
	if(this->indent < 0) {
		this->indent = 0;
	}
}

void CodeWriter::write(const std::string & s)
{
	if(s.empty()) {
		return;
	}

	std::string indentText = this->getIndentText();

	//if(! this->text.empty() && this->text[this->text.size() - 1] == '\n') {
	//	this->text = this->text + indentText;
	//}

	this->text.append(indentText + s);
}

void CodeWriter::writeLine(const std::string & s)
{
	this->write(s + "\n");
}

void CodeWriter::ensureBlankLine()
{
	size_t size = this->text.size();
	int blankLineCount = 0;
	while(size > 0 && this->text[--size] == '\n') {
		++blankLineCount;
	}
	while(blankLineCount < 2) {
		this->text.append("\n");
		++blankLineCount;
	}
}

const std::string & CodeWriter::getText() const
{
	return this->text;
}


} // namespace metagen
